tools/webpack/packages.js

```diff
@@ -143,29 +152,103 @@ module.exports = {
 	performance: {
 		hints: false, // disable warnings about package sizes
 	},
+	optimization: {
+		...baseConfig.optimization,
+		runtimeChunk: {
+			name: 'runtime',
+		},
+	},
 	plugins: [
 		...plugins,
 		new DependencyExtractionWebpackPlugin( { injectPolyfill: true } ),
 		new CopyWebpackPlugin( {
-			patterns: gutenbergPackages
-				.map( ( packageName ) => ( {
-					from: '*.css',
-					context: `./packages/${ packageName }/build-style`,
-					to: `./build/${ packageName }`,
-					transform: stylesTransform,
-					noErrorOnMissing: true,
+			patterns: bundledPackagesPhpConfig.concat(
+				Object.entries( copiedVendors ).map( ( [ to, from ] ) => ( {
+					from: `node_modules/${ from }`,
+					to: `vendors/${ to }`,
 				} ) )
-				.concat( bundledPackagesPhpConfig )
-				.concat(
-					Object.entries( copiedVendors ).map( ( [ to, from ] ) => ( {
-						from: `node_modules/${ from }`,
-						to: `build/vendors/${ to }`,
-					} ) )
-				),
+			),
 		} ),
 		new MomentTimezoneDataPlugin( {
 			startYear: 2000,
 			endYear: 2040,
 		} ),
 	].filter( Boolean ),
 };
+
+if ( baseConfig.mode === 'development' ) {
+	Object.keys( config.entry ).forEach( ( packageName ) => {
+		config.entry[ packageName ].import.unshift(
+			...glob.sync( `./packages/${ packageName }/build-style/*.css` )
+		);
+	} );
+	config.entry[ '🔥hot' ] = {
+		import: [
+			// Runtime code for hot module replacement.
+			'webpack/hot/only-dev-server.js',
+			// Set the log level for HMR to "error".
+			'./tools/webpack/set-hot-log-level.js',
+			// Dev server client for web socket transport, hot and live reload logic.
+			`webpack-dev-server/client/index.js?hot=true&live-reload=false&port=${ WEBPACK_DEV_SERVER_PORT }`,
+		],
+		filename: '🔥hot.js',
+	};
+
+	config.devServer = {
+		static: {
+			directory: resolve( __dirname, '../../build' ),
+			publicPath: '/build/',
+		},
+		liveReload: false,
+		devMiddleware: {
+			writeToDisk( filePath ) {
```

Comment:
That's something we could reuse in the default config shipped with `@wordpress/scripts` as we don't need to save these files on disk that are specific for hot module replacement.

https://github.com/WordPress/gutenberg/blob/31e9078d1209d428dc04b78f0102a7f71c52d928/packages/scripts/config/webpack.config.js#L311-L327

Reply:
In fact, we don't actually **need** to write anything to the disk 😆. The only reason we do now is because the PHP backend uses `glob` to search for the built files on the disk, which I think is probably an anti-pattern we can avoid? Practically we want to avoid the calls for IO since it's often the most costly operations, I guess we can replace that with a generated manifest JSON and/or something like a API call in development mode. I try to keep everything the same in this PR as much as possible to prevent any accidental breakage though 😅.

Reply:
There are also asset files produced (in core it’s a single file, maybe in Gutenberg, too) which work like a manifest. Anyway, if you see some room for improvement you probably should look at optimizing WP core, too.